virttest/utils_misc.py

```diff
@@ -4085,3 +4093,272 @@ def get_model_features(model_name):
+        self.result = CmdResult(command)
+
+        # allow for easy stdin input by string, we'll let subprocess create
+        # a pipe for stdin input and we'll write to it in the wait loop
+        if isinstance(stdin, basestring):
+            self.string_stdin = stdin
+            stdin = subprocess.PIPE
+        else:
+            self.string_stdin = None
+
+        if verbose:
+            logging.debug("Running '%s'" % command)
+        # Ok, bash is nice and everything, but we might face occasions where
+        # it is not available. Just do the right thing and point to /bin/sh.
+        shell = '/bin/bash'
+        if not os.path.isfile(shell):
+            shell = '/bin/sh'
+        self.sp = subprocess.Popen(command, stdout=subprocess.PIPE,
+                                   stderr=subprocess.PIPE,
+                                   preexec_fn=self._reset_sigpipe,
+                                   close_fds=close_fds,
+                                   shell=True,
+                                   executable=shell,
+                                   stdin=stdin)
+
+    def output_prepare(self, stdout_file=None, stderr_file=None):
+        self.stdout_file = stdout_file
+        self.stderr_file = stderr_file
+
+    def process_output(self, stdout=True, final_read=False):
+        """output_prepare must be called prior to calling this"""
+        if stdout:
+            pipe, buf, tee = self.sp.stdout, self.stdout_file, self.stdout_tee
+        else:
+            pipe, buf, tee = self.sp.stderr, self.stderr_file, self.stderr_tee
+
+        if final_read:
+            # read in all the data we can from pipe and then stop
+            data = []
+            while select.select([pipe], [], [], 0)[0]:
+                data.append(os.read(pipe.fileno(), 1024))
+                if len(data[-1]) == 0:
+                    break
+            data = "".join(data)
+        else:
+            # perform a single read
+            data = os.read(pipe.fileno(), 1024)
+        buf.write(data)
+        tee.write(data)
+
+    def cleanup(self):
+        self.stdout_tee.flush()
+        self.stderr_tee.flush()
+        self.sp.stdout.close()
+        self.sp.stderr.close()
+        self.result.stdout = self.stdout_file.getvalue()
+        self.result.stderr = self.stderr_file.getvalue()
+
+    def _reset_sigpipe(self):
+        signal.signal(signal.SIGPIPE, signal.SIG_DFL)
+
+
+class AsyncJob(BgJob):
+
+    def __init__(self, command, stdout_tee=None, stderr_tee=None, verbose=True,
+                 stdin=None, stderr_level=DEFAULT_STDERR_LEVEL, kill_func=None,
+                 close_fds=False):
+        super(AsyncJob, self).__init__(command, stdout_tee=stdout_tee,
+                                       stderr_tee=stderr_tee, verbose=verbose, stdin=stdin,
+                                       stderr_level=stderr_level, close_fds=close_fds)
+
+        # start time for CmdResult
```

Comment:
start/Start

Reply:
Adopt and modify other same spell issue.